enable watcher local testing using k3d (#329)
kyma-project/lifecycle-manager: docs/developer/watcher-local-test.md

# Test watcher locally in two-cluster-setup using K3D

### Contents

### Export customizable attributes
```shell
export KLM_PATH=<local-path-of-kyma-lifecycle-manager-project>
export KLM_VERSION=<docker-image-tag-for-local-testing>
export KUBE_CONFIG_DIR=<path-to-export-local-kubeconfigs-to>
```

### Create KCP cluster
Run the following command to create a local control-plane (KCP) cluster:
```shell
k3d cluster create kcp-local --port 9080:80@loadbalancer \
--registry-create k3d-registry.localhost:0.0.0.0:5111 \
--k3s-arg '--no-deploy=traefik@server:0'
```

### Install Istio and Manifest CRD
Run the following command to install Istio and Manifest CRD on the KCP cluster:
```shell
istioctl install -y && \
kubectl apply -f https://raw.githubusercontent.com/kyma-project/module-manager/main/config/crd/bases/operator.kyma-project.io_manifests.yaml
```

### Build and push lifecycle-manager image
1. Run the following command to get KCP cluster kubeconfig and export it:
```shell

> k3d kubeconfig get kcp-local > $KUBE_CONFIG_DIR/kcp-local.yaml && \

Reviewer: Instead of this we can simply switch `kubectl config use-context ...`

export KUBECONFIG=$KUBE_CONFIG_DIR/kcp-local.yaml
```
2. Run the following command to build and push lifecycle-manager image to local k3d registry:
```shell
export K3D_REG=k3d-registry.localhost:5111 && \
make -C $KLM_PATH docker-build IMG=$K3D_REG/lifecycle-manager:$KLM_VERSION && \

> make -C $KLM_PATH docker-push IMG=$K3D_REG/lifecycle-manager:$KLM_VERSION

Reviewer: these 2 steps are not required. Instead you can use the `latest` tag on KLM gcr image.

Author: I am not sure. the idea is that when the installation implementation changes, this process should be used to test the watcher is still working correctly.

Reviewer 2: I agree with @adityabhatia , since you even build the image yourself with docker-build you can assume latest or any other tag

```

### Deploy lifecycle-manager on KCP:
Run the following command to deploy lifecycle-manager on the KCP cluster:
```shell
export K3D_REG=k3d-registry.localhost:5111 && \

> make -C $KLM_PATH install && \

Reviewer: can we make `make install` a part of `make local-deploy-with-watcher`, just like we use `make deploy`

make -C $KLM_PATH local-deploy-with-watcher IMG=$K3D_REG/lifecycle-manager:$KLM_VERSION
```

### Apply sample module templates for sample-kyma:

> Run the following command to deploy lifecycle-manager on the KCP cluster:

Reviewer: the lifecycle-manger was installed in the previous step. The following command instead adds moduletemplates

```shell
export KUBECONFIG=$KUBE_CONFIG_DIR/kcp-local.yaml && \
kubectl apply -f $KLM_PATH/config/samples/component-integration-installed/operator_v1alpha1_moduletemplate_kcp-module.yaml && \

> kubectl apply -f $KLM_PATH/config/samples/component-integration-installed/operator_v1alpha1_moduletemplate_skr-module.yaml

Reviewer: why do you install `skr-module`. As far as I see below only `kcp-module `is used?

```
### Apply sample Kyma CR
1. Run the following command to create a local kyma-runtime (SKR) cluster:
```shell
k3d cluster create skr-local
```
2. Run the following command to generate and apply sample Kyma CR and its corresponding secret on KCP:
```shell
export KUBECONFIG=$KUBE_CONFIG_DIR/kcp-local.yaml && \
cat << EOF | kubectl apply -f -
---
apiVersion: v1
kind: Secret
metadata:
  name: kyma-sample
  namespace: default
  labels:
    "operator.kyma-project.io/kyma-name": "kyma-sample"
    "operator.kyma-project.io/managed-by": "lifecycle-manager"
data:
  config: $(echo "${$(k3d kubeconfig get skr-local)/0.0.0.0/host.k3d.internal}" | base64)
---
apiVersion: operator.kyma-project.io/v1alpha1
kind: Kyma
metadata:
  name: kyma-sample
  namespace: default
spec:
  channel: regular
  sync:
    enabled: true
  modules:
    - name: kcp-module
EOF
```
### Verify that watcher-webhook is installed on SKR

> By checking the Kyma Object events, verify that the `SKRWebhookIsReady` ready condition has a `True` Status. 

Reviewer: ```suggestion
By checking the `Kyma CR `events, verify that the `SKRWebhookIsReady` ready condition is set to `True`. 
```

```yaml
Status:                                              
    Conditions:                                        
        Message:               skrwebhook is synchronized
        Observed Generation:   1               
        Reason:                SKRWebhookIsReady
        Status:                True
        Type:                  Ready
```
### Edit kyma on SKR

> By adding the following line to the Kyma CR `spec`, trigger the watcher KCP update

Reviewer: provide a command to easliy switch contexts, e.g. `kubectl config use-context k3d-skr-local` ?

# Test watcher locally in two-cluster-setup using K3D

### Contents

### Create KCP cluster
Run the following command to create a local control-plane (KCP) cluster:
```shell
k3d cluster create kcp-local --port 9080:80@loadbalancer \
--registry-create k3d-kyma-registry.localhost:0.0.0.0:5111 \
--k3s-arg '--no-deploy=traefik@server:0'
```

### Install Istio and Manifest CRD
Run the following command to install Istio and Manifest CRD on the KCP cluster:
```shell
kubectl config use-context k3d-kcp-local && \
istioctl install -y && \
kubectl apply -f https://raw.githubusercontent.com/kyma-project/module-manager/main/config/crd/bases/operator.kyma-project.io_manifests.yaml
```

### Build and push lifecycle-manager image
Run the following commands to build and push lifecycle-manager image to local k3d registry:
```shell
export K3D_REG=k3d-kyma-registry.localhost:5111 && \
make docker-build IMG=$K3D_REG/lifecycle-manager:latest && \
make docker-push IMG=$K3D_REG/lifecycle-manager:latest
```

### Deploy lifecycle-manager on KCP:
Run the following command to deploy lifecycle-manager on the KCP cluster:
```shell

> kubectl config use-context k3d-kcp-local && \

Reviewer: You can remove `use-context ` in each command. Rather switch once during the course of the documentation.

export K3D_REG=k3d-kyma-registry.localhost:5111 && \

> make local-deploy-with-watcher IMG=$K3D_REG/lifecycle-manager:latest

Reviewer: For these 2 sections you could use only: 
` make local-deploy-with-watcher IMG=eu.gcr.io/kyma-project/lifecycle-manager:latest`

{"level":"INFO","date":"2022-12-16T10:52:07.98353109Z","caller":"event/skr_events_listener.go:103","msg":"request could not be verified - Event will not be dispatched","context":{"Module":"Listener","error":"empty certificate"}}       
{"level":"INFO","date":"2022-12-16T10:52:13.406499949Z","caller":"controllers/kyma_controller.go:87","msg":"reconciling modules","context":{"controller":"kyma","controllerGroup":"operator.kyma-project.io","controllerKind":"Kyma","kyma":{"name":"kyma-sample","namespace":"default"},"namespace":"default","name":"kyma-sample","reconcileID":"64add714-e3fb-4513-827d-9bb9ba2325a3"}}                                                                                          
{"level":"INFO","date":"2022-12-16T10:52:13.451737025Z","caller":"controllers/kyma_controller.go:206","msg":"syncing state","context":{"controller":"kyma","controllerGroup":"operator.kyma-project.io","controllerKind":"Kyma","kyma":{"name":"kyma-sample","namespace":"default"},"namespace":"default","name":"kyma-sample","reconcileID":"64add714-e3fb-4513-827d-9bb9ba2325a3","state":"Processing"}}
```

> ### verify SKR-watcher logs

Reviewer: Can we combine the verify logs sections and capitalise the first letter

By watching the `skr-webhook` deployment's logs, verify that the KCP request is sent successfully
```log
1.6711877286771238e+09    INFO    skr-webhook    Kyma UPDATE validated from webhook 
1.6711879279507768e+09    INFO    skr-webhook    incoming admission review for: operator.kyma-project.io/v1alpha1, Kind=Kyma 
1.671187927950956e+09    INFO    skr-webhook    KCP    {"url": "http://host.k3d.internal:9080/v1/lifecycle-manager/event"} 
1.6711879280545895e+09    INFO    skr-webhook    sent request to KCP successfully for resource default/kyma-sample 
1.6711879280546305e+09    INFO    skr-webhook    kcp request succeeded
```

> ### cleanup

Reviewer: Capitalise here as well

1. Run the following command to remove the local testing clusters
```shell
k3d cluster rm kcp-local skr-local
```
2. Run the following commands to remove the kubeconfigs of the local testing clusters:
```shell
kubectl config unset users.admin@k3d-skr-cluster && \
kubectl config unset users.admin@k3d-kcp-cluster && \
kubectl config unset clusters.k3d-skr-cluster && \
kubectl config unset clusters.k3d-kcp-cluster && \
kubectl config unset contexts.k3d-skr-cluster && \
kubectl config unset contexts.k3d-kcp-cluster

> ```

Reviewer: this can be removed

# Test watcher locally in two-cluster-setup using K3D

### Contents

### Export customizable attributes
```shell
export KLM_PATH=<local-path-of-kyma-lifecycle-manager-project> && \
export KLM_VERSION=<docker-image-tag-for-local-testing>  && \

> export KUBE_CONFIG_DIR=<path-to-export-local-kubeconfigs-to>

Reviewer 2: Not used anymore I believe

```

### Create KCP cluster
Run the following command to create a local control-plane (KCP) cluster:
```shell
k3d cluster create kcp-local --port 9080:80@loadbalancer \

> --registry-create k3d-registry.localhost:0.0.0.0:5111 \

Reviewer 2: ```suggestion
--registry-create k3d-kyma-registry.localhost:0.0.0.0:5111 \
```
since this is the same registry name as with Kyma CLI. On that note, maybe we could/should switch to that?

--k3s-arg '--no-deploy=traefik@server:0'
```

### Install Istio and Manifest CRD
Run the following command to install Istio and Manifest CRD on the KCP cluster:
```shell
kubectl config use-context k3d-kcp-local && \

> istioctl install -y && \

Reviewer 2: which Istioctl was this tested with? do we need to specify further here?

Reviewer: I think this should be specified

Author: the istioctl version or the steps to install istioctl?

Reviewer: Maybe demo profile and a link the istioctl installation as reference
`istioctl install --set profile=demo`

kubectl config use-context k3d-kcp-local && \
kubectl apply -f $KLM_PATH/config/samples/component-integration-installed/operator_v1alpha1_moduletemplate_kcp-module.yaml && \
kubectl apply -f $KLM_PATH/config/samples/component-integration-installed/operator_v1alpha1_moduletemplate_skr-module.yaml
```
### Apply sample Kyma CR
1. Run the following command to create a local kyma-runtime (SKR) cluster:
```shell
k3d cluster create skr-local
```
2. Run the following command to generate and apply sample Kyma CR and its corresponding secret on KCP:
```shell
kubectl config use-context k3d-kcp-local && \
cat << EOF | kubectl apply -f -
---
apiVersion: v1
kind: Secret
metadata:
  name: kyma-sample
  namespace: default
  labels:
    "operator.kyma-project.io/kyma-name": "kyma-sample"
    "operator.kyma-project.io/managed-by": "lifecycle-manager"
data:
  config: $(echo "${$(k3d kubeconfig get skr-local)/0.0.0.0/host.k3d.internal}" | base64)
---
apiVersion: operator.kyma-project.io/v1alpha1
kind: Kyma
metadata:
  name: kyma-sample
  namespace: default
spec:
  channel: regular
  sync:
    enabled: true
  modules:
    - name: kcp-module
EOF
```
### Verify that watcher-webhook is installed on SKR

> By checking the `Kyma CR` events, verify that the `SKRWebhookIsReady` ready condition is set to `True`.

Reviewer 2: For me this step is never reached. Instead I face this error:

```
│ {"level":"ERROR","date":"2022-12-23T10:30:38.70169913Z","caller":"controller/controller.go:326","msg":"Reconciler error","context":{"controller":"kyma","controllerGroup":"operator.kyma-project.io","c │
│ ontrollerKind":"Kyma","kyma":{"name":"kyma-sample","namespace":"default"},"namespace":"default","name":"kyma-sample","reconcileID":"485ec836-31eb-4cb0-8ac4-7e560cdcf136","error":"failed to verify web │
│ hook resources: unable to create manifest processor: open /skr-webhook/manifest: permission denied"}
```

Thus both kyma CRs are in error state.

Reviewer 2: Also the SKR Condition is not set to false but simply does not exist so I think the Condition is actually not set correctly in the loop.

Author: I have tried to reproduce it but didn't succeed. It would be nice if you could try it again with the latest and verify if the same error is happening again?

```yaml
Status:                                              
    Conditions:                                        
        Message:               skrwebhook is synchronized
        Observed Generation:   1               
        Reason:                SKRWebhookIsReady
        Status:                True
        Type:                  Ready
```
### Edit kyma on SKR
1. Run the following command to switch the context to using the SKR cluster
```shell
kubectl config use-context k3d-skr-local
```
2. Add the following line to the Kyma CR `spec` to trigger the watcher KCP update
```yaml
  modules:
  - name: skr-module
```
### verify lifecycle-manager logs
By watching the lifecycle-manager logs, verify that the listener is logging messages indicating the reception of a message from the watcher

> ```log

Reviewer 2: The log looks much different for me and does not trigger anything:

```
{"level":"INFO","date":"2023-01-03T13:11:11.785053459Z","logger":"kcp-system/klm-watcher","caller":"controllers/watcher_controller.go:69","msg │
│ ":"Reconciliation loop starting","context":{"controller":"watcher","controllerGroup":"operator.kyma-project.io","controllerKind":"Watcher","wa │
│ tcher":{"name":"klm-watcher","namespace":"kcp-system"},"namespace":"kcp-system","name":"klm-watcher","reconcileID":"6c0edf3d-7f25-420b-82aa-b2 │
│ afc99ba9f9"}}                                                                                                                                  │
│ {"level":"INFO","date":"2023-01-03T13:11:31.819069218Z","logger":"kcp-system/klm-watcher","caller":"controllers/watcher_controller.go:69","msg │
│ ":"Reconciliation loop starting","context":{"controller":"watcher","controllerGroup":"operator.kyma-project.io","controllerKind":"Watcher","wa │
│ tcher":{"name":"klm-watcher","namespace":"kcp-system"},"namespace":"kcp-system","name":"klm-watcher","reconcileID":"e19dcb08-864c-44aa-983a-02 │
│ 324f1bc7e7"}}             
```

{"level":"INFO","date":"2022-12-16T10:52:07.98353109Z","caller":"event/skr_events_listener.go:103","msg":"request could not be verified - Event will not be dispatched","context":{"Module":"Listener","error":"empty certificate"}}       
{"level":"INFO","date":"2022-12-16T10:52:13.406499949Z","caller":"controllers/kyma_controller.go:87","msg":"reconciling modules","context":{"controller":"kyma","controllerGroup":"operator.kyma-project.io","controllerKind":"Kyma","kyma":{"name":"kyma-sample","namespace":"default"},"namespace":"default","name":"kyma-sample","reconcileID":"64add714-e3fb-4513-827d-9bb9ba2325a3"}}                                                                                          
{"level":"INFO","date":"2022-12-16T10:52:13.451737025Z","caller":"controllers/kyma_controller.go:206","msg":"syncing state","context":{"controller":"kyma","controllerGroup":"operator.kyma-project.io","controllerKind":"Kyma","kyma":{"name":"kyma-sample","namespace":"default"},"namespace":"default","name":"kyma-sample","reconcileID":"64add714-e3fb-4513-827d-9bb9ba2325a3","state":"Processing"}}
```
### verify SKR-watcher logs

> By watching the `skr-webhook` deployment's logs, verify that the KCP request is sent successfully

Reviewer 2: since your guide does not assume creating a deployment for module manager, the modules are never reconciled and stay stuck in processing with the Kyma CR. thus I was not able to verify this step with the guide.
